torch_xla/csrc/reduction.cpp

```diff
@@ -518,4 +523,67 @@ xla::XlaOp BuildLogsumexp(xla::XlaOp input,
   return logs + max_in_dim;
 }
 
+/**
+ * @brief Returns a random permutation of integers from 0 to n - 1
+
+ * [Algorithm of randperm]: randperm is implemented by sorting an
+ * arange tensor of size n with randomly generated keys. When random
+ * keys are different from each other, all different permutations have
+ * the same probability.
+ *
+ * However, there is a pitfall here:
+ * For better performance, these N random keys are generated independently,
+ * and there is no effort to make sure they are different at the time of
+ * generation. When two keys are identical, stable sorting algorithms will not
+ * permute these two keys. As a result, (0, 1) will appear more often than (1,
+ * 0).
+ *
+ * The PyTorch CUDA implementation for randperm performs some additional
+ * post-processing to resolve this pitfall, which guarantees that the
+ * permutation is truely uniformaly distributed. In contrast, this XLA
+ * implementation does not guarantee a true uniform distribution. However,
+ * it performs multiple rounds of sorting with random keys. Since the
+ * probability of multiple successive key collision is small, the resulting
+ * distribution of the permutations approximates a uniform distribution.
+ *
+ * TODO: Implement post-processing for the generated random permutation to
+ * resolve this pitfall, i.e. ensuring that the permutations are truly
+ * randomly uniformly distributed.
+ * Link to PyTorch CUDA implementation of this post-processing:
+ *
+ https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/Randperm.cu
+ */
+xla::XlaOp BuildRandpermOut(int64_t n, xla::XlaBuilder* builder) {
```

Comment:
@thisisalbertliang hi Albert, regarding this algorithm's implementation, can you point to me the one that you referenced?

Reply:
@thisisalbertliang 
Please help @vanbasten23 pick up from where you left things off. If you can offer additional context beyond his question to get started, it'd be awesome. Hope you are well :)

Reply:
Sure thing! Thanks @vanbasten23 for picking up on this, and hope you are well too @miladm .

My implementation of `BuildRandpermOut` (i.e. this function that you commented on) is largely the same as the `BuildRandperm` function that (for some reason) got deleted in Feb 2020: https://github.com/pytorch/xla/commit/237d32629f0f7d9d385c7289b1e2c746c8dfb26d

With this implementation, I remember I was passing all the test cases in `test_aten_xla_tensor.cpp`, including `AtenXlaTensorTest.TestRandperm`. However, I was failing some other tests. Specifically, they are: [failed_tests_log.txt](https://github.com/pytorch/xla/files/8600708/failed_tests_log.txt). I got stuck on debugging these tests as I wasn't sure how they are related to `randperm_out`.

@vanbasten23 Feel free to completely replace my `RandpermOut` implementation in this PR with your own one. I recommend taking a look at the `randperm` code that got deleted in Feb 2020 (https://github.com/pytorch/xla/commit/237d32629f0f7d9d385c7289b1e2c746c8dfb26d), since that implementation seems to have worked in the past.

Side Note: not sure if it helps, but PyTorch CUDA also has its own implementation of `Randperm`, which you can reference here: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/Randperm.cu. Their implementation, however, seems more complicated.